### DISCUSSION TITLE
Importing a score in PDF format into Audiveris, resolution greatly goes down and program is unable to recognize most of the music.

### DISCUSSION THREAD

**Discussion by SeoKyungAhn** [Audiveris/audiveris#700] [Q&A]
Hello. Apologies in advance if I do not provide all the necessary information in this first post as I am not very familiar with OMR/Github in general.

I have downloaded Audiveris and am attempting to scan a score that I purchased in PDF form (For non-commercial, academic use).

The PDF is of very high resolution when I access it through my Firefox Browser (at least, I think, because when I zoom in, the image doesn't get grainy at all).

However, when I imported it into Audiveris, the image became very grainy compared to the original on both the Gray and the Binary. 

When I tried to transcribe, It gave me this error for ALL the pages, even for the ones that have staves and music:

> With an interline value of 9 pixels, either this sheet contains no staves, or the picture resolution is too low (try 300 DPI). Okay for discarding this sheet?

To troubleshoot, I tried increasing contrast technique using GIMP as suggested by the handbook, at 500 DPI (didn't want to go too high as suggested by the tutorial).

The result was that now the program was able to recognize the key signature better, as well as most of the time signatures, but it did not pick up on any of the notes, nor time signatures in the middle of the sheet.

I saw that there was a similar issue with imported PDFs being very low resolution in Audiveris (Issue #162 ), but I did not see a clear-cut fix (Again, apologies if I missed a solution that was posted).

What could be causing this issue? I have attached relevant information below:

The PDF in question, original:

[CivilizationVI_America_TheIndustrialEra.pdf](https://github.com/Audiveris/audiveris/files/13398517/CivilizationVI_America_TheIndustrialEra.pdf)

The PDF, after increasing contrast with GIMP at 500 DPI:

[CivilizationVI_America_TheIndustrialEra_Contrasted.pdf](https://github.com/Audiveris/audiveris/files/13398417/CivilizationVI_America_TheIndustrialEra_Contrasted.pdf)

This is what shows up on the CMD when I first launch the program:

```
INFO  []                       CLI 282  | CLI args: []
INFO  []              TesseractOCR 116  | TESSDATA_PREFIX value: null
INFO  []                      Main 402  | Environment:
- Audiveris:    5.3.1:5aa4d06dc
- OS:           Windows 10 10.0
- Architecture: amd64
- Java VM:      Java HotSpot(TM) 64-Bit Server VM (build 21.0.1+12-LTS-29, mixed mode, sharing)
- OCR Engine:   Tesseract OCR, version 5.3.1
INFO  []             AliasPatterns 134  | Alias patterns: [(IMSLP[0-9]*)-.*]
INFO  []                   MainGui 573  | Audiveris version 5.3.1
INFO  []                   MainGui 574  |
LogUtil. Property logback.configurationFile not defined, skipped.
LogUtil. No C:\Users\Brian\AppData\Roaming\AudiverisLtd\audiveris\config\logback.xml, skipped.
LogUtil. Configuration found jar:file:/C:/Program%20Files/Audiveris/lib/audiveris.jar!/res/logback.xml
LogUtil. Logging to C:\Users\Brian\AppData\Roaming\AudiverisLtd\audiveris\log\20231117T160929.log

INFO  []            PluginsManager 210  | No C:\Users\Brian\AppData\Roaming\AudiverisLtd\audiveris\config\plugins.xml file found
INFO  []           BasicClassifier 277  | Classifier loaded XML norms.
INFO  []        AbstractClassifier 396  | Classifier data loaded from default uri jar:file:/C:/Program%20Files/Audiveris/lib/audiveris.jar!/res/basic-classifier.zip
INFO  []                  Versions 301  | Versions. Poll frequency: Weekly, next poll on: 24-Nov-2023
```

This is the error log for the last page:

```
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] Loaded image 15 2549x3299 from C:\Users\Brian\OneDrive\OneDrive - California Institute of Technology\Sheet Music\Civ6 America Industrial\CivilizationVI_America_TheIndustrialEra_Contrasted.pdf
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] CivilizationVI_America_TheIndustrialEra_Contrasted#15 With an interline value of 9 pixels, either this sheet contains no staves, or the picture resolution is too low (try 300 DPI).
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] Beam  guessed height:  4 -- 0.50 of 9 interline
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] Beam measured height:  8 -- 0.83 of [3..9] range at 25% of needed quorum
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] No reliable beam height found, guessed value: 4
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] Scale{ interline(8,9,9) line(1,2,3) beam(4 extra)}
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] Global slope: 0.00000
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] Retrieved line clusters: 24 of sizes [5] with interline(8,9,9)
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] Systems: #1[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13] #2[14] #3[15, 16, 17, 18, 19, 20, 21, 22, 23, 24]
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] No brace partner at level:2 for Peak{370-379 F#4491 T#1 [BRACE]}
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] No brace partner at level:2 for Peak{370-378 F#4492 T#2 [BRACE]}
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] No brace partner at level:2 for Peak{368-378 F#4495 T#4 [BRACE, BRACE_BOTTOM]}
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] No brace partner at level:2 for Peak{368-379 F#4498 T#6 [BRACE, BRACE_BOTTOM]}
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] No brace partner at level:2 for Peak{367-379 F#4499 T#7 [BRACE, BRACE_BOTTOM]}
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] No brace partner at level:2 for Peak{368-379 F#4502 T#9 [BRACE, BRACE_BOTTOM]}
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] No brace partner at level:2 for Peak{370-379 F#4503 T#10 [BRACE]}
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] System#1
[CivilizationVI_America_TheIndustrialEra_Contrasted#15]    PartGroup{bracket number:1 barline:true staves:1-5}
[CivilizationVI_America_TheIndustrialEra_Contrasted#15]    PartGroup{bracket number:1 barline:true staves:6-12}
[CivilizationVI_America_TheIndustrialEra_Contrasted#15]    PartGroup{bracket number:1 barline:false staves:13}
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] System#3
[CivilizationVI_America_TheIndustrialEra_Contrasted#15]    PartGroup{bracket number:1 barline:true staves:19-24}
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] Indentation detected for system#2
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] 2 pages found in sheet
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] Page #1: 13 parts along 1 system
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] Page #2: 8 parts along 2 systems
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] Staff#6 no header clef.
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] Discarding alpha band ...
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] Converting max RGB to gray
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] stem(2 max:3)
[CivilizationVI_America_TheIndustrialEra_Contrasted#15] Error in performing [LOAD, BINARY, SCALE, GRID, HEADERS, STEM_SEEDS, BEAMS, LEDGERS, HEADS, STEMS, REDUCTION, CUE_BEAMS, TEXTS, MEASURES, CHORDS, CURVES, SYMBOLS, LINKS, RHYTHMS, PAGE] java.util.concurrent.ExecutionException: java.lang.RuntimeException: org.audiveris.omr.math.Line$UndefinedLineException: Not enough defining points : 1
[CivilizationVI_America_TheIndustrialEra_Contrasted] PartCollation needed in {Score 1}
[CivilizationVI_America_TheIndustrialEra_Contrasted] Could not transcribe book java.lang.NullPointerException: Cannot invoke "org.audiveris.omr.sheet.rhythm.Measure.getVoices()" because the return value of "org.audiveris.omr.sheet.Part.getFirstMeasure()" is null
```

This is what the last page looks like after the transcribe feature is used. This is representative of basically all the pages: None of the staves are recognizes, nor are the notes/rests/etc.

![image](https://github.com/Audiveris/audiveris/assets/77988239/ad2018e0-7ac4-4161-90b7-7fdfb2d33870)

**Comment by hbitteur** [comment]
@SeoKyungAhn 
I just tried your input file with Audiveris latest version.
The first sheet with some concrete music content is the sheet number  4.
Running the SCALE step on this sheet gives an interline value of 9 pixels.

For Audiveris engine, which works best with interlines around 20 pixels, this value is way too small, leading the engine to suspect there are no music staves on this sheet. Hence the warning dialog you got.

Unfortunately, this 9-pixel interline is correct. It is the consequence of a very large score printed on a standard sheet.
What you could try is to feed Audiveris with a better scaled input.

Here is what I did:
I used Foxit Reader (a freeware on Windows) and loaded your input .pdf file. Then I asked Foxit Reader to "print" this document, not to a real printer, but to a virtual PDF (Foxit Reader PDF Printer), for which I selected a target page size of A3 (double the size of A4).
In fact I printed all the pages except the first 3 pages which contain no music.
And then I got a new .pdf file which I called A3-CivilizationVI_America_TheIndustrialEra.pdf. Here it is:
[A3-CivilizationVI_America_TheIndustrialEra.pdf](https://github.com/Audiveris/audiveris/files/13399906/A3-CivilizationVI_America_TheIndustrialEra.pdf)

This new PDF file exhibits a staff interline of 13 pixels. Audiveris engine accepts this value. 
I noticed just two problems:
- A few note heads are not correctly recognized. I suspect the scale is still a bit too low, impeding a correct template matching for note heads. I tried to use an A2 print format, this resulted in monstruous pages of 35 millions pixels, which the engine could not swallow...
- The 1-line percussion staves, though declared as such in the book parameters, are not detected. I need to investigate this further.

**Comment by hbitteur** [comment]
Playing with your example, I encountered other problems related to some of the 1-line staves.
I will have to strenghen the checks to discard many false candidates.

Anyway, there are a few cases where the same physical staff goes from 5-line to 1-line, or the other way around.

See the following example:
![image](https://github.com/Audiveris/audiveris/assets/22053634/39dc8030-aaf4-4de9-973e-ffeb0225ec5a)

Or this one, on another sheet:
![image](https://github.com/Audiveris/audiveris/assets/22053634/520fc386-42d1-44db-b413-e0a59f888d12)

**Reply by SeoKyungAhn** [reply]
Thank you hbitteur and apologies for not getting back to this thread sooner. In that case would it be advisable to split the scores into two halves (for example, top and bottom), and then re-size them to a standard sheet (A4 or A3)?

Also, regarding your comment:

"I suspect the scale is still a bit too low, impeding a correct template matching for note heads"

Does this mean, in layman's terms that I would need to increase the resolution? Or are you saying that the interline value is still too small?

I would appreciate it if you could provide me with what you feel like would be the most efficient workflow is to address this issue.

Thank you.

**Comment by manderpe** [comment]
This sounds, in some ways, similar to the  problem that I am having with a PDF input file taken, most likely, from some score editing software.  I have used Audiveris quite successfully on scanned sheet music but when I opened this PDF file Audiveris failed to interpret many bars.  The problem, I suspect, is that the original PDF is a vector graphic file rather than raster graphic which explains why it can be scaled significantly in a PDF viewer without loss of detail.  Possibly when Audiveris reads the file it interprets it to be at the lowest resolution - I don't know.  Also, in my case, a file size of 1.4MB for a 70 plus page score is a dead giveaway.  I tried printing to paper and scanning with a good scanner at 400 dpi and that gave me much better results.  So, I tried using ghostscript to size the PDF directly using something like:

gs -dNOPAUSE -dBATCH -sDEVICE=tiffg4 \
   -r400 \
   -sOutputFile=output.tif \
   input.pdf

Note that the resultant file  is a tiff file.  Better results but still it missed many bars.  So, assuming that Audiveris was trained on imperfect notes and stems I added a little Gaussian blur using:

magick output_600dpi.tif -blur 0x0.5 output_600dpi_soft.tif

Better but still nowhere near the printed and scanned version.  Even adding a little noise using:

magick test_5pages.tif -attenuate 0.03 +noise Gaussian test_noise.tif

No better.  In my case the score is the pianist's copy of a piano trio and the missing bars were mostly in the violin and cello and they are scaled to 70% so this suggested that Audiveris might be struggling with the size of the notes so I tried increasing the pagesize to 150%  using:

 gs -dNOPAUSE -dBATCH -sDEVICE=tiffgray \
   -r400 \
   -dDEVICEWIDTHPOINTS=1386 -dDEVICEHEIGHTPOINTS=1800 \
   -dFIXEDMEDIA -dPDFFitPage \
   -sOutputFile=output_scaled.tif \
   input.pdf

Absolutely no difference so the size of the notes on the page are not the problem it appears.  Maybe the music font used in the original is just not recognised by Audiveris.  I just find it fascinating that I have better luck with old IMSLP sheet music scanned on a scanner than I do on directly saved PDF files from music notation software.   

In summary:

Source file:  https://s9.imslp.org/files/imglnks/usimg/0/0d/IMSLP1016195-PMLP1457433-FARRENC_Trio_no._2_en_r%C3%A9_mineur_op._34_-_Conducteur.pdf

- Vector PDF rasterized via Ghostscript at 400 DPI greyscale
- Same score printed to paper and scanned at 400 DPI on a Fujitsu scanner produces significantly better recognition
- Scaling up 150% and applying Gaussian blur make no difference
- The failing bars are consistently the same ones across all attempts
- The failing staves are the reduced-size violin and cello parts in a conductor score

My next step is to save another known Musescore score as a PDF and run it through Audiveris and see whether it has troubles with the vector graphics or not.
